src/mcelog.c

```diff
@@ -0,0 +1,431 @@
+  vl.values_len = 1;
+  vl.time = cdtime();
+
+  sstrncpy(vl.plugin, MCELOG_PLUGIN, sizeof(vl.plugin));
+  sstrncpy(vl.type, "errors", sizeof(vl.type));
+  if (strlen(mr.dimm_name) > 0) {
+    snprintf(plugin_instance, DATA_MAX_NAME_LEN, "%s_%s", mr.location,
+             mr.dimm_name);
+    sstrncpy(vl.plugin_instance, plugin_instance, sizeof(vl.plugin_instance));
+  } else
+    sstrncpy(vl.plugin_instance, mr.location, sizeof(vl.plugin_instance));
+
+  sstrncpy(vl.type_instance, "corrected_memory_errors",
+           sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.corrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  snprintf(type_instance, DATA_MAX_NAME_LEN, "corrected_memory_errors_in_%s",
+           mr.corrected_err_timed_period);
+  sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.corrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  sstrncpy(vl.type_instance, "uncorrected_memory_errors",
+           sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.uncorrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  snprintf(type_instance, DATA_MAX_NAME_LEN, "uncorrected_memory_errors_in_%s",
+           mr.uncorrected_err_timed_period);
+  sstrncpy(vl.type_instance, type_instance, sizeof(vl.type_instance));
+  vl.values = &(value_t){.derive = (derive_t)mr.uncorrected_err_total};
+  plugin_dispatch_values(&vl);
+
+  return 0;
+}
+
+static void *poll_worker(__attribute__((unused)) void *arg) {
+  char buf[DATA_MAX_NAME_LEN] = {0};
+  mcelog_memory_rec_t memory_record;
+  _Bool received_data = 0;
+  FILE *p_file;
+
+  struct pollfd poll_fd = {
+      .fd = g_mcelog_config.sock_fd, .events = POLLIN | POLLPRI,
+  };
+
+  while (1) {
+
+    pthread_mutex_lock(&mcelog_thread_lock);
+    if (mcelog_thread_shutdown == 1) {
+      pthread_mutex_unlock(&mcelog_thread_lock);
+      pthread_exit(NULL);
+    }
+    pthread_mutex_unlock(&mcelog_thread_lock);
+    int poll_ret = poll(&poll_fd, 1, POLL_TIMEOUT);
+
+    if (poll_ret > 0) {
+      if (poll_fd.revents & POLLNVAL) {
+        if (connect_to_client_socket() != 0)
+          usleep(POLL_TIMEOUT);
+      }
+      if ((poll_fd.revents & POLLERR) || (poll_fd.revents & POLLHUP)) {
+        /* connection is broken */
+        shutdown(g_mcelog_config.sock_fd, 0);
+        ERROR("%s: Connection to socket is broken", MCELOG_PLUGIN);
+        notification_t n = {
+            NOTIF_FAILURE, cdtime(), "", "", MCELOG_PLUGIN, "", "", "", NULL};
+        ssnprintf(n.message, sizeof(n.message),
+                  "Connection to mcelog socket is broken.");
+        sstrncpy(n.type_instance, "mcelog_status", sizeof(n.type_instance));
+        mcelog_dispatch_notification(n);
+        break;
+      }
+      if ((poll_fd.revents & POLLIN) || (poll_fd.revents & POLLPRI)) {
+        if ((p_file = fdopen(dup(g_mcelog_config.sock_fd), "r")) != NULL) {
```

Comment:
More very long blocks, deeply nested. Please handle the error, not the positive case and remove levels of indentation. The parsing should probably go into its own function to make it more readable.